apicast/src/plugin.lua

```diff
@@ -0,0 +1,57 @@
+-- Apicast provides a plugin system. To add a custom plugin, we only need to
+-- define a module that implements a few methods like rewrite(), access(), etc.
+-- that are called in the appropriate nginx phases.
+-- The goal of this lua module is to implement the methods that are not likely
+-- to be changed from plugin to plugin so we do not have to repeated code
+-- across all of them.
+
+local balancer = require 'balancer'
+
+-- Plugins need access to provider and configuration, so we need to expose them
+local _M = { provider = require 'provider',
```

Comment:
`configuration` does not need to be exposed. I realized it wasn't needed, but forgot to delete that part.

Regarding inheriting from Apicast itself, the problem is that xc does not need post_action(), body_filter(), nor header_filter(). Also, there are parts of access() that should be common to both of them, but not all.